contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.10;
+import "../libraries/Address.sol";
+
+// types
+import "../types/BaseAllocator.sol";
+
+// interfaces
+import "../interfaces/IMasterChef.sol";
+import "../interfaces/ISushiBar.sol";
+
+/**
+ *  Contract deploys liquidity from treasury into the Onsen program,
+ *  earning $SUSHI that can be staked and/or deposited into the treasury.
+ */
+contract OnsenAllocatorV2 is BaseAllocator {
+    using SafeERC20 for IERC20;
+
+    /* ========== STATE VARIABLES ========== */
+
+    address immutable sushi; // $SUSHI token
+    address immutable xSushi; // $xSUSHI token
+    address immutable masterChef; // Onsen contract
+
+    address immutable treasury; // Olympus Treasury
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        address _treasury,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+        require(_treasury != address(0));
+        treasury = _treasury;
+
+        // approve for safety, sushi is being staked and xsushi is being instantly sent to treasury and that is fine
+        // but to be absolutely safe this one approve won't hurt
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
```

Comment:
the masterchef id is going to be a different number to ids of the deposits in the `TreasuryExtender` right?

Reply:
@Paul4912 sorry but not following, I am using the id that is passed to the deposit function as the id for the deposit into masterchef
is that the id you are referring to?

Reply:
So for the masterchef contract each LP token farm is mapped to an ID. To do anything like deposit you need to input the correct ID that corresponds to the LP token. 

The id in _update(id) is the id corresponding to the deposit id that the allocator and treasury extender references. This is not the same id as the id that we need to put into the masterchef contract made by sushi.

As PARABRAHMAN0 mentioned in another comment you should probably have an array of integers the same length as IERC20[] internal _tokens to store the masterchef id's of all your LP tokens. This way when you construct the _tokens in the constructor you can also construct another array that maps 1 to 1 to easily access their masterchef id's. You won't need a new mapping since it will be sharing the mapping(uint256 => uint256) public tokenIds mapping.

Reply:
Finally I understood how to get the id of the Onsen Pool based on the LP token, I am keeping locally a mapping to easily get the Onsen pool id based on the LP token.
This logic was updated.